### ISSUE TITLE
[AutoDiff] Differentiability analysis is too strict on differentiable function with one non-differentiable argument

### ISSUE THREAD

**Issue by BradLarson** [apple/swift#68774]
**Description**

In the following example, an error is raised for `f()` not being differentiable for the reason that you cannot differentiate through a non-differentiable result:

```swift
import _Differentiation

@differentiable(reverse)
func f(_ x: Double) -> Double {
    return ifs([x < 0.0, x > 0.0], [3.0 * x, 5.0 * x], 7.0 * x)
}

@differentiable(reverse, wrt: (values, defaultValue))
public func ifs(_ conditions: [Bool], _ values: [Double], _ defaultValue: Double) -> Double {
    let count = withoutDerivative(at: min(values.count, conditions.count))
    for i in 0 ..< count {
        if conditions[i] { return values[i] }
    }
    return defaultValue
}
```

However, the `ifs()` function is differentiable, and `f()` can be made differentiable in the following manner:

```swift
@differentiable(reverse)
func f(_ x: Double) -> Double {
    let localValue = withoutDerivative(at: x)
    return ifs([localValue < 0.0, localValue > 0.0], [3.0 * x, 5.0 * x], 7.0 * x) // desired implementation
}
```

This seems to indicate that the differentiability analysis is being confused by the fact that `x` is a Differentiable value and active in the function.

**Steps to reproduce**

Place the first example above in a standalone file and attempt to build.

**Expected behavior**

The example should build without any errors about differentiability.

**Environment**

This requires a Swift.org compiler with the _Differentiation module present. Verified on the 2023-09-20 nightly trunk snapshot toolchain, but this same differentiability analysis fails on trunk snapshots going back into the middle of 2022, so I don't believe it's a recent change.

**Comment by asl** [comment]
> This seems to indicate that the differentiability analysis is being confused by the fact that x is a Differentiable value and active in the function.

@BradLarson well, but `x` is really active by definition. It's a function input and it affects the output.

**Comment by jkshtj** [comment]
Posting some more details regarding what's exactly causing an issue here.

`x` is indeed active, which the compiler determines correctly. What's failing here is the computation of the first arg of `ifs` - the boolean array, in `f`. The SIL for that looks something like this -

```
%r1 = 0.0
%r2 = x < %r1
%r3 = x > %r1
%r4 = [%r2, %r3]
// During activity analysis r4 is marked useful and
// eventually this "usefulness" is progagated to %r2 
// and %r3, which are not differentiable since `float.<`
// and `floaf.>` are not differentiable. 
%r5 = apply ifs (%r4, ... )
```

**Comment by jkshtj** [comment]
Are there issues in not marking arguments as 'useful' if the `apply` is not differentiable with respect to it?

**Comment by asl** [comment]
> Are there issues in not marking arguments as 'useful' if the `apply` is not differentiable with respect to it?

That's the other way round: we must differentiate a function wrt all active arguments. Why? This is... well, by definition of derivative (this is a bit of simplification here, but for active values we know that they are not fixed and that the derivative wrt them is non-zero). 

Doing what you are asking will effectively kill the majority of non-differentiable diagnostics – we will silently ignore non-derivative stuff.

**Comment by jkshtj** [comment]
Could you elaborate your explanation with an example? 

I'm not sure I see how, if we only mark certain SSA values as not-useful (or don't mark them as useful) it would affect the diagnostics in the case when a does not differentiate w.r.t all active arguments?

**Comment by asl** [comment]
> Could you elaborate your explanation with an example?
> 
> I'm not sure I see how, if we only mark certain SSA values as not-useful (or don't mark them as useful), it would affect the diagnostics in the case when a function does not differentiate w.r.t all active arguments?

I believe we already discussed this. The properties useful / varied / active has nothing to do with the differentiation at the first place. It's the properties of the input program and its values. However, as a consequence to this, if the value is useful, we know that the derivative of the function wrt this value is non-zero. And here is a big difference between "zero" and "does not exist". Replacing blindly latter with former is a mistake.

**Comment by jkshtj** [comment]
> I believe we already discussed this. 
Yeah sorry for repeating the same point. I started wondering if there is gap in how we decide whether or not to differentiate a SIL instruction because in the case of this specific example, we attempted to differentiate and only got a failure because `Double.<` is not differentiable. But let's say the signature for `ifs` was `([Double], [Double], Double) -> Double` instead of `([Bool], [Double], Double) -> Double`, we might have calculated `f`'s derivative incorrectly. 

**However**, I just checked the differentiation for an `apply` and it only calculates adjoints of input args wrt which the function is differentiable. So, the problem that I was thinking of does not exist.
